src/batch.rs

```diff
@@ -235,6 +235,15 @@ pub fn verify_batch(
         }.into());
     }
 
+    // If we have been given a trivial batch size, we can exit early
+    // without going through the trouble of batch verification.
+    if signatures.len() == 0 {
+        return Ok(());
+    } else if signatures.len() == 1 {
+        use crate::ed25519::signature::Verifier;
+        return public_keys[0].verify(messages[0], &signatures[0]);
```

Comment:
Hmm. Thinking out loud here. I'm not entirely sure if I'm okay with mixing the verification equations because of the subtle differences between the two (but anyone should feel free to try to convince me). See for example [my and @valerini's comments on this issue](https://github.com/dalek-cryptography/ed25519-dalek/pull/117). I suppose in general the single verification is safer than the linearised batching formula, but I still worry that mixing the two behaviours will result in difficult to debug verification errors like the following situation:

1) I call `verify_batch()` on bad signature `(s1,R1)`, which fails.
2) I call `verify_batch()` on bad signature `(s1,R1)` and crafted signature `(s2,R2)` which probabilistically cancels out the linearisation factors and succeeds.

However, I suppose the above behaviour is better than the current behaviour where calling `verify_batch()` on `(s2,R2)` alone would probabilistically succeed.

Okay, I think I've convinced myself that this behaviour is safer. My only remaining concern is the debugging confusions it might cause with a signature not verifying if passed into this function by itself versus in a batch with N>1.

Reply:
Hi @isislovecruft  -  I am afraid you have the wrong person, probably looking for @kchalkias instead

Reply:
Table 3 in [the paper](https://eprint.iacr.org/2020/1244) details the issue at hand:
![Selection_097](https://user-images.githubusercontent.com/4142/97455032-c678b200-190d-11eb-89e2-53271ac1a855.png)

Dalek has a cofactorless single verification with a cofactored batch verification (column "[2]+[3]"), so we can create what you'd call "false negatives" from the PoV of batch verification as a heuristic: a set S of signatures that will (probabilistically) pass batch verification but fail (iterated) single verification.

The present PR eliminates this possibility when N = |S| = 1 by making sure only single verification applies, but does not change the case |S| > 1, as you mentioned.

One way to fix the issue is to:
1. add a cofactored single signature verification,
2. make sure the batch verification is only available when using this cofactored verification by uniting them both under a "cofactored" feautre flag (which would replace the current "batch" feature) 
3. then implement the spirit of this PR: reduce the batch verification call when N = 1 to the (now cofactored) single verification.

I'm happy to help with 1. & 2.